bench/micro/Bench/Database/LSMTree/Monoidal.hs

```diff
@@ -0,0 +1,288 @@
+{-# LANGUAGE OverloadedLists #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Bench.Database.LSMTree.Monoidal (benchmarks) where
+
+import           Control.DeepSeq
+import           Criterion.Main
+import qualified Data.Vector as V
+import           Data.Void
+import           Data.Word
+import           Database.LSMTree.Extras
+import           Database.LSMTree.Extras.Orphans ()
+import           Database.LSMTree.Internal.Serialise
+import           Database.LSMTree.Monoidal (resolveDeserialised)
+import qualified Database.LSMTree.Monoidal as Monoidal
+import qualified Database.LSMTree.Normal as Normal
+import           GHC.Generics
+import           Prelude hiding (getContents)
+import           System.Directory (removeDirectoryRecursive)
+import qualified System.FS.API as FS
+import qualified System.FS.BlockIO.API as FS
+import qualified System.FS.BlockIO.IO as FS
+import qualified System.FS.IO as FS
+import           System.IO.Temp
+import           System.Random
+
+benchmarks :: Benchmark
+benchmarks = bgroup "Bench.Database.LSMTree.Monoidal" [
+      benchInsertsVsMupserts
+    , benchLookupsInsertsVsMupserts
+    , benchNormalResolveVsMonoidalResolve
+    ]
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+newtype K = K Word64
+  deriving stock Generic
+  deriving newtype (Show, Eq, Ord, Num, NFData, SerialiseKey)
+  deriving anyclass Uniform
```

Comment:
That doesn't work -- the `uniformM` function can't be coerced because of the monad `m`

https://hackage.haskell.org/package/random-1.2.1.2/docs/System-Random-Stateful.html#v:uniformM

Reply:
Ah, TIL.